internal/juju/jaas.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package juju
+
+import (
+	"github.com/canonical/jimm-go-sdk/v3/api"
+	"github.com/canonical/jimm-go-sdk/v3/api/params"
+	jujuapi "github.com/juju/juju/api"
+
+	"github.com/juju/errors"
+)
+
+type jaasClient struct {
+	SharedClient
+	getJaasApiClient func(jujuapi.Connection) JaasAPIClient
+}
+
+func newJaasClient(sc SharedClient) *jaasClient {
+	return &jaasClient{
+		SharedClient: sc,
+		getJaasApiClient: func(conn jujuapi.Connection) JaasAPIClient {
+			return api.NewClient(conn)
+		},
+	}
+}
+
+func (jc *jaasClient) AddRelations(tuples []params.RelationshipTuple) error {
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+	cl := jc.getJaasApiClient(conn)
+	req := params.AddRelationRequest{
+		Tuples: tuples,
+	}
+	return cl.AddRelation(&req)
+}
+
+func (jc *jaasClient) DeleteRelations(tuples []params.RelationshipTuple) error {
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+	cl := jc.getJaasApiClient(conn)
+	req := params.RemoveRelationRequest{
+		Tuples: tuples,
+	}
+	return cl.RemoveRelation(&req)
```

Comment:
what error is returned if the relation does not exists? i believe we should not return an error in that case..

Reply:
Same discussion as above, these APIs are not idempotent so it will fail. And again, difficult to make idempotent because they accept an array of objects.